redis/cluster.py

```diff
@@ -0,0 +1,1884 @@
+                    "A ``db`` querystring option can only be 0 in cluster mode"
+                )
+            kwargs.update(url_options)
+            startup_nodes.append(ClusterNode(kwargs['host'], kwargs['port']))
+        elif host is not None and port is not None:
+            startup_nodes.append(ClusterNode(host, port))
+        elif len(startup_nodes) == 0:
+            # No startup node was provided
+            raise RedisClusterException(
+                "RedisCluster requires at least one node to discover the "
+                "cluster. Please provide one of the followings:\n"
+                "1. host and port, for example:\n"
+                " RedisCluster(host='localhost', port=6379)\n"
+                "2. list of startup nodes, for example:\n"
+                " RedisCluster(startup_nodes=[ClusterNode('localhost', 6379),"
+                " ClusterNode('localhost', 6378)])")
+
+        # Update the connection arguments
+        # Whenever a new connection is established, RedisCluster's on_connect
+        # method should be run
+        # If the user passed on_connect function we'll save it and run it
+        # inside the RedisCluster.on_connect() function
+        self.user_on_connect_func = kwargs.pop("redis_connect_func", None)
+        kwargs.update({"redis_connect_func": self.on_connect})
+        kwargs = cleanup_kwargs(**kwargs)
+
+        self.encoder = Encoder(
+            kwargs.get("encoding", "utf-8"),
+            kwargs.get("encoding_errors", "strict"),
+            kwargs.get("decode_responses", False),
+        )
+        self.cluster_error_retry_attempts = cluster_error_retry_attempts
+        self.command_flags = self.__class__.COMMAND_FLAGS.copy()
+        self.node_flags = self.__class__.NODE_FLAGS.copy()
+        self.debug_mode = debug
+        self.read_from_replicas = read_from_replicas
+        self.reinitialize_counter = 0
+        self.reinitialize_steps = reinitialize_steps
+        self.nodes_manager = None
+        self.nodes_manager = NodesManager(
+            startup_nodes=startup_nodes,
+            from_url=from_url,
+            require_full_coverage=require_full_coverage,
+            skip_full_coverage_check=skip_full_coverage_check,
+            **kwargs,
+        )
+
+        self.cluster_response_callbacks = CaseInsensitiveDict(
+            self.__class__.CLUSTER_COMMANDS_RESPONSE_CALLBACKS)
+        self.result_callbacks = CaseInsensitiveDict(
+            self.__class__.RESULT_CALLBACKS)
+        self.commands_parser = CommandsParser(self)
+        self._lock = threading.Lock()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def __del__(self):
+        self.close()
+
+    def disconnect_connection_pools(self):
+        for node in self.get_nodes():
+            if node.redis_connection:
+                node.redis_connection.connection_pool.disconnect()
```

Comment:
Further into the disconnect code I see that we pass on OSError, but there could be client disconnects here - since this is the furthest things are exposed.

Reply:
added a try except